tests: added two unit test applications

RanRegev opened the pull request:
This PR adds two test targets that check
the conversions between vs and 1905
for beacon metrics query and response.
it is used during development and can also
be used during regualr builds as a test.
currently the DUT, the conversion cpp function
implementation is compiled twice:
* as part of the son-slave
* as part of the test application
The best would be to create a library for conversion.

Signed-off-by: Ran regev <ran.regev@devalore.com>

arnout commented:
You should also add it to CI. Should be sufficient to add the following to .gitlab-ci.yml:
```
1905_beacon_query_to_vs_test:
    extends: .run-test-in-docker
```

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+#include <bcl/network/network_utils.h>
+#include <tlvf/CmduMessageTx.h>
+
+constexpr const int RET_OK          = 0;
```
`constexpr` implies `const`, so no need to use both. `constexpr` is enough and preferred.
According to naming convention, the name of the variable should use lower_snake_case (UPPER_SNAKE_CASE is for #define)

RanRegev replied:
Will fix.
And yes, you are correct: http://eel.is/c++draft/dcl.spec#dcl.constexpr-10

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+std::ostream &operator<<(std::ostream &o, const beerocks_message::sBeaconRequest11k &request)
+{
+    // NOTE: not all fields are printed
+    // TODO: complete all fields
```
(nitpick) 
When I add a TODO, there's always someone who asks me in the PR comments to create an issue and add the issue number here.

RanRegev replied:
good advice

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+      << "ssid:                     " << request.ssid << "\n"
+      << "use_optional_ap_ch_report:" << +request.use_optional_ap_ch_report << "\n";
+
+    for (int idx = 0; idx < request.use_optional_ap_ch_report; ++idx) {
```
(nitpick) 
Use size_t instead of int

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+        o << "channel[" << idx << "]:    " << std::dec << +request.ap_ch_report[idx] << "\n";
+    }
+
+    return o;
```
I'd add a std::flush to ostream before returning

RanRegev replied:
ok, will do. I'll use std::endl instead of "\n":
https://en.cppreference.com/w/cpp/io/manip/endl
http://eel.is/c++draft/ostream.manip#2

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    // TODO: complete all fields
+
+    o << "\n"
+      << "measurement_mode:         " << +request.measurement_mode << "\n"
```
I think the '+' sign before the variable is a typo.
Be aware it also appears in some other lines below

RanRegev replied:
Take a look at ["Method D"](https://bytefreaks.net/programming-2/c/c-how-to-print-an-unsigned-character-unsigned-byte-uint8_t-using-cout)
and also at [unary _ operator](http://eel.is/c++draft/expr.unary.op#7), [conv.prom](http://eel.is/c++draft/conv.prom#1), [conv.rank](http://eel.is/c++draft/conv.rank#1.3) and finally on [cout printing arithmetics](http://eel.is/c++draft/ostream.inserters.arithmetic).

In words: the type of `unary + operator` on `unsigned char` is `int`. This is because the `unsigned char` was promoted to `int`. The promotion to int is because its `integer conversion rank` is less than of `int`.

mariomaz replied:
TIL

arnout replied:
Nice! I should have read this yesterday, then I had a learning to report on the sprint meeting!

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    // currently we don't print anything except for the params field
+    // we might want to print / verify the action op (currently hard coded in the class)
+    o << "ACTION_MONITOR_CLIENT_BEACON_11K_REQUEST:"
+      << "\n-----------------------------------------\n"
```
Some static analyzers would complain because you are using the same literal string more than once.

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    return o;
+}
+
+void fillTxValues(ieee1905_1::CmduMessageTx &tx)
```
Class members and functions shall use lower_snake_case.

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    // escape I
+    auto tlv = tx.getClass<wfa_map::tlvBeaconMetricsQuery>();
+    if (!tlv) {
+        LOG(ERROR) << "this Tx message does not contain tlv beacon metrics query";
```
You'd probably want to return in case of error

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+
+    // fill some values
+    tlv->associated_sta_mac() = sMacAddr{0x01, 0x02, 0x03, 0x04, 0x05, 0x06};
+    tlv->operating_class()    = 0x73;
```
Some static analyzers would complain because you are using magic numbers here

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    // preperation for beerocks
+
+    // class list
+    constexpr const uint16_t BUF_SIZE_BEEROCKS = 10000;
```
constexpr implies const.
Use lower_snake_case for the variable name

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    constexpr const uint16_t BUF_SIZE_1905_TX = 10000;
+    uint8_t buff_1905_tx[BUF_SIZE_1905_TX];
+    ieee1905_1::CmduMessageTx tx(buff_1905_tx, BUF_SIZE_1905_TX);
+    tx.create(0x1, ieee1905_1::eMessageType::BEACON_METRICS_QUERY_MESSAGE);
```
Check returned value

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    // DUT -
+    gate::load(/*dst*/ beerocks, /*src*/ rx);
```
I think `gate` and `load` are not meaningful names as they don't reflect what the method is really doing

RanRegev replied:
[take a look](https://github.com/prplfoundation/prplMesh/pull/1273#discussion_r424333678)

mariomaz replied:
I see I'm not the only one who thinks it's not a good name, then it's a smell you should think it twice.
Some people say that comments in code should not be necessary if code was clear enough.
What name will you choose when creating a gate between other two worlds?

mariomaz commented on agent/src/beerocks/slave/test/1905_beacon_query_to_vs_test.cpp:
```diff
@@ -0,0 +1,131 @@
+
+    // verification (hunman readable output_
+    beerocks_message::cACTION_MONITOR_CLIENT_BEACON_11K_REQUEST &msg = *beerocks;
+    std::cout << msg;
```
Unit tests should be designed to run without interaction, no human should be required to decide if tests PASS or FAIL by inspecting console output.
A better approach is to use assertions which automatically indicate test results.

RanRegev replied:
ok

mariomaz reviewed (changes requested):
Use assertions instead of print statements to verify the code executed correctly.

Regarding dual compilation, it's ok if code under test has to be compiled twice. You cannot even avoid it if you create a library. Most of the times is simply because your target platform is not the same as your testing platform. But even if they were, reason is compilation flags must be different for production binary than for unit tests. For example, you might want to use some optimization flags to create production binary that you have to avoid if you want to generate an accurate code coverage report. Also you need special compilation flags to generate code coverage information that you should skip in production binary.

RanRegev commented:
> Use assertions instead of print statements to verify the code executed correctly.
> 
> Regarding dual compilation, it's ok if code under test has to be compiled twice. You cannot even avoid it if you create a library. Most of the times is simply because your target platform is not the same as your testing platform. But even if they were, reason is compilation flags must be different for production binary than for unit tests. For example, you might want to use some optimization flags to create production binary that you have to avoid if you want to generate an accurate code coverage report. Also you need special compilation flags to generate code coverage information that you should skip in production binary.

I think that tests that are changing the DUT to their need are only partial tests.
If you are compiling _differently_ for production and for testing then something will eventually slip away. As for different targets, yes, you should compile differently for each target - this is inherent in the word "target". Library _does_ solve the problem as you compile the library (for each target) once and link against it with your _code_ and with your _test_.

RanRegev committed:
tests: gate unit test: added google test

Added google tests to check gate::load functionality
1905->vs: beacon query.
vs->1905: beacon response.
Added unit_tests directory and changed cmake accordingly

Signed-off-by: Ran Regev <ran.regev@devalore.com>

mariomaz commented on agent/src/beerocks/slave/CMakeLists.txt:
```diff
@@ -34,3 +34,33 @@ target_link_libraries(${PROJECT_NAME} rt dl bcl btl tlvf elpp btlvf bwl bpl ${LI
+if (BUILD_TESTS)
+    set(TEST_PROJECT_NAME ${PROJECT_NAME}_unit_tests)
+    set(unit_tests_sources
+        gate/unit_test/gate_test.cpp
```
Directory name for unit tests in other projects has been named `unit_tests`

RanRegev replied:
I checked other directories... missed that `s`.
Fixed.

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    // class list
+
+    constexpr const uint16_t BUF_SIZE_BEEROCKS = 10000; // just an arbitrary size
```
No need for `const` after `constexpr`
UPPER_CASE is for #define's

RanRegev replied:
done.

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    constexpr const uint16_t BUF_SIZE_1905_TX = 10000; // just an arbitrary size
+    uint8_t buff_1905_tx[BUF_SIZE_1905_TX];
+    ieee1905_1::CmduMessageTx cmdu_1905(buff_1905_tx, BUF_SIZE_1905_TX);
+    cmdu_1905.create(0x1, ieee1905_1::eMessageType::BEACON_METRICS_QUERY_MESSAGE);
```
Please check return value

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    // Rx - built from the tx message above
+    ieee1905_1::CmduMessageRx rx(cmdu_1905.getMessageBuff(), cmdu_1905.getMessageBuffLength());
+    rx.parse();
```
Please check return value

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    ASSERT_NE(tlv_becon_metrics_query, nullptr);
+
+    // DUT -
+    beerocks::gate::load(/*dst*/ beerocks, /*src*/ rx);
```
Have I ever mentioned that `gate` and `load` are not appropriate names ;-)
You had to add comments in the parameters to remember what is being loaded into what

RanRegev replied:
I prefer `beerocks << rx;` or `rx >> beerocks;` but it leaves no space for future parameters if needed (like some context to do the translation): `beerocks::gate::load(beerocks, rx, context_to_support_translation)`

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    auto &beerocks_params = beerocks->params();
+
+    // set
+    ASSERT_EQ(beerocks_params.channel, channel_number);
```
Please consider using EXPECT_EQ instead of ASSERT_EQ
See [Assertions](https://github.com/google/googletest/blob/master/googletest/docs/primer.md#assertions)

RanRegev replied:
done.

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -0,0 +1,121 @@
+    ClassList cl(buff_beerocks, BUF_SIZE_BEEROCKS);
+
+    // beerocks
+    cl.addClass<beerocks_message::cACTION_MONITOR_CLIENT_BEACON_11K_REQUEST>();
```
Please check return value

mariomaz commented on agent/src/beerocks/slave/gate/unit_test/gate_test.cpp:
```diff
@@ -118,4 +122,53 @@ TEST(gate_beacon_query_test, loading_vs_specific_from_1905_beacon_query)
+    ClassList cl(buff_beerocks, BUF_SIZE_BEEROCKS);
+
+    // beerocks
+    cl.addClass<beerocks_message::cACTION_MONITOR_CLIENT_BEACON_11K_RESPONSE>();
```
Same comments here than in previous test about checking returned values and constexpr + const

mariomaz reviewed (approved):
Some very simple to fix comments, so approving

arnout reviewed (approved):
The unit test isn't executed in CI, so it's quite useless. Look at the bcl_unit_test in .gitlab-ci.yml for how it's done.

RanRegev committed:
tests: added beerocks_agent_unit_tests to ci

Signed-off-by: Ran Regev <ran.regev@devalore.com>
